### PR TITLE
Feature/m0081 meja 154 introduce feature code chunks

### PR DESCRIPTION
**Pull request by frivieremicroej**

### CODE REVIEW

File: VEEPortingGuide/architecture8Migration.rst [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+As a reminder, refer to the :ref:`Architecture 8.1.0 Changelog <changelog-8.1.0>` section for the complete list of changes and updates.
+
+.. _architecture8_migration_llkernel:
+
+Migrate Your LLKERNEL Implementation
+------------------------------------
+
+This section only applies if your LLKERNEL was based on legacy :ref:`In-Place Installation mode <feature_inplace_installation>`. 
+The following code is a ``LLKERNEL_impl.c`` template for migrating your current implementation using the new APIs.
+The code logic based on a ``malloc/free`` implementation does not need to be changed.
```
Comment (jerome-leroux) [inline VEEPortingGuide/architecture8Migration.rst:17]: I would add something like this (I am not sure I totally understood the changes):
The Kernel Working Buffer no longer exists, and the Features can no longer be allocated in this buffer. Memory allocation for the Features will now use the function `LLKERNEL_IMPL_allocateFeature()`. The functions `LLKERNEL_IMPL_allocateWorkingBuffer()` and `LLKERNEL_IMPL_freeWorkingBuffer()` have been removed.

File: VEEPortingGuide/multiSandbox.rst [hunk 2]
```diff
@@ -106,95 +101,51 @@ A Feature ``.fo`` file is composed of the following elements:
...
+   :alt: Feature Installation Steps
    :align: center
    :scale: 70%
 
-   Custom Feature Installation Steps
+   Feature Installation Steps
 
 The Abstraction Layer implementation is responsible for providing the following elements:
 
 - the address location where the Feature will be installed,
```
Comment (jerome-leroux) [inline VEEPortingGuide/multiSandbox.rst:129]:
```suggestion
- the location where the Feature will be installed,
```

**Review by jerome-leroux — CHANGES_REQUESTED** [review]

**Comment by frivieremicroej** [comment]
https://docs.microej.com/en/feature-m0081meja-154_introduce_feature_code_chunks/

File: VEEPortingGuide/architecture8Migration.rst [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+   void LLKERNEL_IMPL_freeFeature(int32_t handle) {
+      struct installed_feature* f = (struct installed_feature*)handle;
+      KERNEL_FREE(f->ROM_area);
+      KERNEL_FREE(f->RAM_area);
+      KERNEL_FREE(handle);
+   }
+
+   int32_t LLKERNEL_IMPL_getAllocatedFeaturesCount(void) {
+      // No persistency support
+      return 0;
```
[the remark this answers is not in the document]
Comment (frivieremicroej) [reply VEEPortingGuide/architecture8Migration.rst:56]: I rewrote theses LLAPI, here is the latest version.
```
/**
	 * Gets the number of Features that have been installed into memory from a prior execution of this Kernel.
	 * <p>
	 * This function is called once at the start of the Core Engine. Afterward, the Core Engine will retrieve Feature
	 * handles by calling {@link #getFeatureHandle(int)} for each index in the range <code>[0..allocated_features_count[</code>.
	 *
	 * @return the number of Features that have been installed into memory from a prior execution of this Kernel, <code>0</code> if none.
	 * @see #getFeatureHandle(int)
	 */
	public int getAllocatedFeaturesCount();
	
	/**
	 * Gets the handle of a Feature that has been installed into memory from a prior execution of this Kernel.
	 * <p>
	 * This function is called at the start of the Core Engine for each index from <code>0</code> to
	 * <code>{@link IDynamicFeaturesManager#getAllocatedFeaturesCount()}-1</code>.
	 *
	 * @param index
	 *            an index in the range <code>[0..{@link IDynamicFeaturesManager#getAllocatedFeaturesCount()}[</code>
	 * @return a handle that uniquely identifies the Feature located at the specified index, which has been installed into memory from a prior execution of this Kernel.  (<code>0</code> is no a valid handle).
	 * @see #getAllocatedFeaturesCount()
	 */
	public int getFeatureHandle(int index);
```

File: VEEPortingGuide/architecture8Migration.rst [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+   struct installed_feature{
+      void* ROM_area;
+      void* RAM_area;
+   };
+
+   int32_t LLKERNEL_IMPL_allocateFeature(int32_t size_ROM, int32_t size_RAM) {
+      struct installed_feature* f = (struct installed_feature*)KERNEL_MALLOC(sizeof(struct installed_feature));
+      f->ROM_area = KERNEL_MALLOC(size_ROM);
+      f->RAM_area = KERNEL_MALLOC(size_RAM);
```
[the remark this answers is not in the document]
Comment (frivieremicroej) [reply VEEPortingGuide/architecture8Migration.rst]: done

File: VEEPortingGuide/architecture8Migration.rst [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+   // Your implementation of 'ASSERT(0)'
+   #define KERNEL_ASSERT_FAIL() while(1)
+
+   struct installed_feature{
+      void* ROM_area;
+      void* RAM_area;
+   };
+
+   int32_t LLKERNEL_IMPL_allocateFeature(int32_t size_ROM, int32_t size_RAM) {
+      struct installed_feature* f = (struct installed_feature*)KERNEL_MALLOC(sizeof(struct installed_feature));
```
[the remark this answers is not in the document]
Comment (frivieremicroej) [reply VEEPortingGuide/architecture8Migration.rst]: done

File: VEEPortingGuide/architecture8Migration.rst (hunk 1 above)
Comment (frivieremicroej) [reply VEEPortingGuide/architecture8Migration.rst:20]: OK, most of of you suggestion added to give more details

File: VEEPortingGuide/architecture8Migration.rst [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+   // Your implementation of free()
+   #define KERNEL_FREE(addr) free((void*)addr)
+
+   // Your implementation of 'ASSERT(0)'
+   #define KERNEL_ASSERT_FAIL() while(1)
+
+   // Utility macros for allocating RAM and ROM areas with required alignment constraints
+   #define KERNEL_AREA_GET_MAX_SIZE(size, alignment) (size+(alignment-1))
+   #define KERNEL_AREA_GET_START_ADDRESS(addr, alignment) ((void*)((((int32_t)addr)+alignment-1)&~(alignment-1)))
```
Comment (jerome-leroux) [inline VEEPortingGuide/architecture8Migration.rst:40]:
```suggestion
   #define KERNEL_AREA_GET_MAX_SIZE(size, alignment) ((size)+((alignment)-1))
   #define KERNEL_AREA_GET_START_ADDRESS(addr, alignment) ((void*)((((int32_t)(addr))+(alignment)-1)&~((alignment)-1)))
```

File: VEEPortingGuide/multiSandbox.rst [hunk 7]
```diff
@@ -106,102 +101,61 @@ A Feature ``.fo`` file is composed of the following elements:
 
    Feature ``.fo`` File Content
 
-The ``LLKERNEL_impl.h`` Abstraction Layer interface provides Low Level APIs for allocating and transferring Feature content in different memory areas.
+Feature installation flow allows to install Features in any byte-addressable memory (including ROM) with a minimum amount of required RAM.
```
Comment (jerome-leroux) [inline VEEPortingGuide/multiSandbox.rst:104]: I am not sure about the "byte-addressable" term. One can argue that an external serial flash is "byte-addressable" with a driver.

We should probably use the term "byte-addressable and mapped in the address space of the CPU"?

**Review by jerome-leroux — CHANGES_REQUESTED** [review]
See previous comments

File: VEEPortingGuide/multiSandbox.rst (hunk 7 above)
Comment (frivieremicroej) [reply VEEPortingGuide/multiSandbox.rst]: ok, updated 9b84365aad789f2809770f649c267973e15811cf

File: VEEPortingGuide/architectureChangelog.rst [hunk 8]
```diff
@@ -56,7 +72,7 @@ This major Architecture version update introduces the following main features:
   - On different Kernel Applications provided some conditions are met. 
     Basically, a ``.fo`` built on Kernel 1 can be installed on Kernel 2 if the exposed Kernel APIs are left unchanged.
     See :ref:`feature_portability_control` for more details.
-- Redesigned Feature installation flow. A Feature can now be installed in any byte-addressable memory, including ROM.
+- Redesigned Feature installation flow. A Feature can now be installed in any any byte-addressable memory mapped to the CPU's address space, including ROM.
```
Comment (jerome-leroux) [inline VEEPortingGuide/architectureChangelog.rst:75]:
```suggestion
- Redesigned Feature installation flow. A Feature can now be installed in any byte-addressable memory mapped to the CPU's address space, including ROM.
```

**Review by jerome-leroux — APPROVED** [review]

**Pull request merged**